File: src/main/java/com/theonlytails/ruby/enchantments/StingerEnchantment.java
package com.theonlytails.ruby.enchantments;

import com.theonlytails.ruby.RubyMod;
import com.theonlytails.ruby.registries.EnchantmentRegistry;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.enchantment.EnchantmentType;
import net.minecraft.enchantment.Enchantments;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.inventory.EquipmentSlotType;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.potion.EffectInstance;
import net.minecraft.potion.Effects;
import net.minecraftforge.event.entity.player.AttackEntityEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;
import org.jetbrains.annotations.NotNull;

public class StingerEnchantment extends Enchantment {
    public StingerEnchantment() {
        super(Rarity.VERY_RARE,
                EnchantmentType.WEAPON,
                new EquipmentSlotType[] {EquipmentSlotType.MAINHAND});
    }

    @Override
    public int getMaxLevel() {
        return 2;
    }

    @Override
    public boolean canApply(ItemStack stack) {
        return stack.getItem() == Items.SHEARS || super.canApply(stack);
    }

    @Override
    public boolean canVillagerTrade() {
        return false;
    }

    @Override
    public boolean canApplyTogether(@NotNull Enchantment enchant) {
        return super.canApplyTogether(enchant) &&
               enchant != Enchantments.SHARPNESS &&
               enchant != Enchantments.MENDING;
    }

    @Mod.EventBusSubscriber(modid = RubyMod.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE)
    private static class PoisonedBladeEquipped {
        @SubscribeEvent
        public static void damageWithEnchant(AttackEntityEvent event) {
            PlayerEntity player = event.getPlayer();

            ItemStack heldItemMainhand = player.getHeldItemMainhand();

            Enchantment enchant = EnchantmentRegistry.STINGER.get();

            int enchantLevel = EnchantmentHelper.getEnchantmentLevel(enchant, heldItemMainhand);

            if (EnchantmentHelper.getEnchantments(heldItemMainhand).containsKey(enchant)) {
                if (event.getTarget() instanceof LivingEntity) {
                    LivingEntity target = (LivingEntity) event.getTarget();

                    target.addPotionEffect(new EffectInstance(Effects.POISON, 100, enchantLevel));
                }
            }
        }
    }
}

File: src/main/kotlin/com/github/theonlytails/rubymod/enchantments/StingerEnchantment.kt
package com.github.theonlytails.rubymod.enchantments

import com.github.theonlytails.rubymod.RubyMod
import com.github.theonlytails.rubymod.registries.EnchantmentRegistry
import net.minecraft.enchantment.*
import net.minecraft.entity.LivingEntity
import net.minecraft.inventory.EquipmentSlotType
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.potion.EffectInstance
import net.minecraft.potion.Effects
import net.minecraftforge.event.entity.player.AttackEntityEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod.EventBusSubscriber

class StingerEnchantment : Enchantment(Rarity.VERY_RARE,
	EnchantmentType.WEAPON, arrayOf(EquipmentSlotType.MAINHAND)) {

	override fun getMaxLevel() = 2

	override fun canApply(stack: ItemStack): Boolean {
		return stack.item === Items.SHEARS || super.canApply(stack)
	}

	override fun canVillagerTrade() = false

	public override fun canApplyTogether(enchant: Enchantment): Boolean {
		return super.canApplyTogether(enchant) && enchant !== Enchantments.SHARPNESS && enchant !== Enchantments.MENDING
	}

	@EventBusSubscriber(modid = RubyMod.MOD_ID, bus = EventBusSubscriber.Bus.FORGE)
	private object PoisonedBladeEquipped {
		@SubscribeEvent
		fun damageWithEnchant(event: AttackEntityEvent) {
			val player = event.player
			val heldItemMainhand = player.heldItemMainhand
			val enchant = EnchantmentRegistry.STINGER
			val enchantLevel = EnchantmentHelper.getEnchantmentLevel(enchant, heldItemMainhand)

			if (EnchantmentHelper.getEnchantments(heldItemMainhand).containsKey(enchant)) {
				if (event.target is LivingEntity) {
					val target = event.target as LivingEntity
					target.addPotionEffect(EffectInstance(Effects.POISON, 100, enchantLevel))
				}
			}
		}
	}
}